c/extras/ftrl.cc

```diff
@@ -363,129 +347,139 @@ DataTable* FtrlModel::get_model() {
 
 
 /*
-*  Set an FTRL model, assuming all the validation is done in `py_ftrl.cc`
+*  Other getters and setters.
+*  Here we assume that all the validation for setters is done in `py_ftrl.cc`.
 */
-void FtrlModel::set_model(DataTable* dt_model_in) {
-  dt_model = dtptr(dt_model_in->copy());
+std::vector<uint64_t> Ftrl::get_colnames_hashes() {
+  return colnames_hashes;
 }
 
+size_t Ftrl::get_n_features() {
+  return n_features;
+}
 
-/*
-*  Other getters and setters.
-*  Here we assume that all the validation is done in `py_ftrl.cc`.
-*/
-double FtrlModel::get_a() {
-  return fmp.a;
+
+double Ftrl::get_alpha() {
+  return params.alpha;
+}
+
+
+double Ftrl::get_beta() {
+  return params.beta;
 }
 
 
-double FtrlModel::get_b() {
-  return fmp.b;
+double Ftrl::get_lambda1() {
+  return params.lambda1;
 }
 
 
-double FtrlModel::get_l1() {
-  return fmp.l1;
+double Ftrl::get_lambda2() {
+  return params.lambda2;
 }
 
 
-double FtrlModel::get_l2() {
-  return fmp.l2;
+uint64_t Ftrl::get_d() {
+  return params.d;
 }
 
 
-uint64_t FtrlModel::get_d() {
-  return fmp.d;
+bool Ftrl::get_inter() {
+  return params.inter;
 }
 
 
-bool FtrlModel::get_inter() {
-  return fmp.inter;
+unsigned int Ftrl::get_hash_type() {
+  return params.hash_type;
 }
 
 
-unsigned int FtrlModel::get_hash_type() {
-  return fmp.hash_type;
+unsigned int Ftrl::get_seed() {
+  return params.seed;
 }
 
 
-unsigned int FtrlModel::get_seed() {
-  return fmp.seed;
+size_t Ftrl::get_n_epochs() {
+  return params.n_epochs;
 }
 
 
-size_t FtrlModel::get_n_epochs() {
-  return fmp.n_epochs;
+/*
+*  Set an FTRL model, assuming all the validation is done in `py_ftrl.cc`
+*/
+void Ftrl::set_model(DataTable* dt_model_in) {
+  dt_model = dtptr(dt_model_in->copy());
+  model_trained = true;
 }
 
 
-void FtrlModel::set_a(double a_in) {
-  if (fmp.a != a_in) {
-    fmp.a = a_in;
+void Ftrl::set_alpha(double a_in) {
+  if (params.alpha != a_in) {
+    params.alpha = a_in;
     init_model();
```

Comment:
Why doing `init_model()` here? This is expensive (esp. if you need to set multiple parameters), and probably unnecessary.

Reply:
Because I don't know what will happen when one restarts from a model that was created based on a different `alpha` (also `beta`, `lambda1`, etc).

Reply:
The "expensive" part refers to the fact that when setting the model's parameters, you're calling `set_alpha()`, `set_beta()`, etc -- and each of them does `init_model()` over and over again.

The "unnecessary" comment is less clear. On one hand, we can prohibit changing the model's parameters altogether once it has been trained. Or alternatively, allow the parameters to be changed, and then start training with the existing model data. It is up to the user whether to reset the data or not. However, I feel that force-clearing the data without the user's consent is the worst choice in this case.

Reply:
I agree on the first part. My comment was more about why we may want to discard the trained model, when some parameters are changed.

Yes, when a user changes the individual parameters one by one, it now does unnecessary job of cleaning the model again, and again. I've been actually thinking about changing it to something like this:

- since now one can use `params` to set all the input parameters at once, we may not actually need all the individual getters and setters;
- we can keep the model state when the parameters change, because there is always a `reset_model()` method to initialize the model;
- we can add an additional argument to `set_params` that will control whether or not the model must be initialized, this argument could be `false` by default. In this case it will be up to the user to decide what to do with the trained model.